crates/storage/src/streams.rs

```diff
@@ -0,0 +1,197 @@
+use std::time::Duration;
+
+#[cfg(test)]
+mod tests;
+
+/// Notify that there are new contracts or blocks.
+#[derive(Clone)]
+pub struct Notify {
+    contracts: tokio::sync::watch::Sender<()>,
+    blocks: tokio::sync::watch::Sender<()>,
+}
+
+/// Wait for new data.
+#[derive(Clone)]
+pub struct NewData(tokio::sync::watch::Receiver<()>);
+
+/// State of the stream.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct StreamState {
+    state: State,
+    start: Start,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum State {
+    Pos(Pos),
+    Done,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Pos {
+    page: usize,
+    index: usize,
+}
+
+/// Get data from this point.
+#[derive(Default, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct GetData {
+    /// Page number to get data from.
+    pub page: usize,
+    /// Time to get data from.
+    pub time: Option<Duration>,
+    /// Number to get data from.
+    pub number: Option<u64>,
+}
+
+#[derive(Default, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Start {
+    time: Option<Duration>,
+    number: Option<u64>,
+}
+
+/// Get the next data in the stream.
+pub async fn next_data<F, Fut, D>(
+    get_data: F,
```

Comment:
Minor nit, but I think the general convention is to receive single closure arguments in the last arg (e.g. `fold(0, closure)`)